Launch XML failed to parse value in `set_use_sim_time` tag

### Generated by Generative AI

_No response_

### Operating System:

```Linux 7bf22331d729 5.15.167.4-microsoft-standard-WSL2 #1 SMP Tue Nov 5 00:21:55 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux```

### ROS version or commit hash:

humble

### RMW implementation (if applicable):

rmw_fastrtps_cpp

### RMW Configuration (if applicable):

_No response_

### Client library (if applicable):

_No response_

### 'ros2 doctor --report' output

<details><summary>ros2 doctor --report</summary>

```console

   NETWORK CONFIGURATION
inet         : 127.0.0.1
inet4        : ['127.0.0.1']
inet6        : ['::1']
netmask      : 255.0.0.0
device       : lo
flags        : 73<RUNNING,LOOPBACK,UP>
mtu          : 65536
inet         : 172.22.0.2
inet4        : ['172.22.0.2']
ether        : 02:42:ac:16:00:02
netmask      : 255.255.0.0
device       : eth0
flags        : 4163<BROADCAST,RUNNING,MULTICAST,UP>
mtu          : 1500
broadcast    : 172.22.255.255

   PACKAGE VERSIONS
nav_msgs                                  : latest=4.9.0, local=4.9.0
ament_cmake_test                          : latest=1.3.12, local=1.3.11
ament_cmake_python                        : latest=1.3.12, local=1.3.11
ament_cmake_export_link_flags             : latest=1.3.12, local=1.3.11
rclcpp                                    : latest=16.0.15, local=16.0.12
ros2multicast                             : latest=0.18.14, local=0.18.12
ros2cli_common_extensions                 : latest=0.1.1, local=0.1.1
rcpputils                                 : latest=2.4.5, local=2.4.5
actionlib_msgs                            : latest=4.9.0, local=4.9.0
launch                                    : latest=1.0.10, local=1.0.9
ros2lifecycle                             : latest=0.18.14, local=0.18.12
ament_cmake_gtest                         : latest=1.3.12, local=1.3.11
diagnostic_msgs                           : latest=4.9.0, local=4.9.0
action_msgs                               : latest=1.2.2, local=1.2.1
ros2cli                                   : latest=0.18.14, local=0.18.12
ros_workspace                             : latest=1.0.2, local=1.0.2
ament_cmake_export_include_directories    : latest=1.3.12, local=1.3.11
ament_cppcheck                            : latest=0.12.14, local=0.12.12
sensor_msgs                               : latest=4.9.0, local=4.9.0
launch_yaml                               : latest=1.0.10, local=1.0.9
rosidl_typesupport_fastrtps_cpp           : latest=2.2.2, local=2.2.2
rosidl_cmake                              : latest=3.1.7, local=3.1.6
ament_cmake_export_interfaces             : latest=1.3.12, local=1.3.11
rosidl_runtime_py                         : latest=0.9.3, local=0.9.3
ament_cpplint                             : latest=0.12.14, local=0.12.12
ros_core                                  : latest=0.10.0, local=0.10.0
libyaml_vendor                            : latest=1.2.2, local=1.2.2
ament_cmake_version                       : latest=1.3.12, local=1.3.11
class_loader                              : latest=2.2.0, local=2.2.0
ament_flake8                              : latest=0.12.14, local=0.12.12
geometry_msgs                             : latest=4.9.0, local=4.9.0
rcutils                                   : latest=5.1.7, local=5.1.6
shape_msgs                                : latest=4.9.0, local=4.9.0
rmw_implementation                        : latest=2.8.5, local=2.8.4
rosidl_typesupport_cpp                    : latest=2.0.2, local=2.0.2
ament_cmake_auto                          : latest=1.3.12, local=1.3.11
ament_lint_common                         : latest=0.12.14, local=0.12.12
ament_cmake                               : latest=1.3.12, local=1.3.11
ament_cmake_cppcheck                      : latest=0.12.14, local=0.12.12
trajectory_msgs                           : latest=4.9.0, local=4.9.0
libstatistics_collector                   : latest=1.3.4, local=1.3.4
ament_cmake_gen_version_h                 : latest=1.3.12, local=1.3.11
launch_testing                            : latest=1.0.10, local=1.0.9
ros2interface                             : latest=0.18.14, local=0.18.12
ament_index_python                        : latest=1.4.0, local=1.4.0
ament_cmake_copyright                     : latest=0.12.14, local=0.12.12
ament_cmake_ros                           : latest=0.10.0, local=0.10.0
rosidl_generator_cpp                      : latest=3.1.7, local=3.1.6
launch_testing_ament_cmake                : latest=1.0.10, local=1.0.9
composition_interfaces                    : latest=1.2.2, local=1.2.1
ros2service                               : latest=0.18.14, local=0.18.12
sros2                                     : latest=0.10.7, local=0.10.6
python_cmake_module                       : latest=0.10.0, local=0.10.0
ros2component                             : latest=0.18.14, local=0.18.12
rosidl_cli                                : latest=3.1.7, local=3.1.6
ament_cmake_export_definitions            : latest=1.3.12, local=1.3.11
uncrustify_vendor                         : latest=2.0.2, local=2.0.2
ament_cmake_libraries                     : latest=1.3.12, local=1.3.11
ament_lint                                : latest=0.12.14, local=0.12.12
rosidl_typesupport_fastrtps_c             : latest=2.2.2, local=2.2.2
ament_cmake_export_libraries              : latest=1.3.12, local=1.3.11
ament_index_cpp                           : latest=1.4.0, local=1.4.0
rosidl_typesupport_introspection_cpp      : latest=3.1.7, local=3.1.6
statistics_msgs                           : latest=1.2.2, local=1.2.1
ament_package                             : latest=0.14.1, local=0.14.1
rosidl_runtime_c                          : latest=3.1.7, local=3.1.6
rmw_dds_common                            : latest=1.6.0, local=1.6.0
tinyxml2_vendor                           : latest=0.7.6, local=0.7.6
rcl_logging_spdlog                        : latest=2.3.1, local=2.3.1
ros2doctor                                : latest=0.18.14, local=0.18.12
ament_uncrustify                          : latest=0.12.14, local=0.12.12
launch_ros                                : latest=0.19.10, local=0.19.9
visualization_msgs                        : latest=4.9.0, local=4.9.0
rclcpp_components                         : latest=16.0.15, local=16.0.12
ament_xmllint                             : latest=0.12.14, local=0.12.12
rosidl_parser                             : latest=3.1.7, local=3.1.6
rosidl_typesupport_c                      : latest=2.0.2, local=2.0.2
ament_cmake_gmock                         : latest=1.3.12, local=1.3.11
unique_identifier_msgs                    : latest=2.2.1, local=2.2.1
ros2action                                : latest=0.18.14, local=0.18.12
ament_cmake_include_directories           : latest=1.3.12, local=1.3.11
tracetools                                : latest=4.1.1, local=4.1.1
ament_cmake_lint_cmake                    : latest=0.12.14, local=0.12.12
sros2_cmake                               : latest=0.10.7, local=0.10.6
stereo_msgs                               : latest=4.9.0, local=4.9.0
spdlog_vendor                             : latest=1.3.1, local=1.3.1
rclcpp_lifecycle                          : latest=16.0.15, local=16.0.12
rcl_lifecycle                             : latest=5.3.10, local=5.3.9
rosgraph_msgs                             : latest=1.2.2, local=1.2.1
domain_coordinator                        : latest=0.10.0, local=0.10.0
rosidl_typesupport_introspection_c        : latest=3.1.7, local=3.1.6
rosidl_generator_c                        : latest=3.1.7, local=3.1.6
launch_xml                                : latest=1.0.10, local=1.0.9
ros2launch                                : latest=0.19.10, local=0.19.9
launch_testing_ros                        : latest=0.19.10, local=0.19.9
ament_cmake_flake8                        : latest=0.12.14, local=0.12.12
ros_environment                           : latest=3.2.2, local=3.2.2
rosidl_default_runtime                    : latest=1.2.0, local=1.2.0
rosidl_runtime_cpp                        : latest=3.1.7, local=3.1.6
rcl_interfaces                            : latest=1.2.2, local=1.2.1
ros2node                                  : latest=0.18.14, local=0.18.12
ament_cmake_target_dependencies           : latest=1.3.12, local=1.3.11
rosidl_default_generators                 : latest=1.2.0, local=1.2.0
ament_cmake_cpplint                       : latest=0.12.14, local=0.12.12
ament_lint_auto                           : latest=0.12.14, local=0.12.12
ament_copyright                           : latest=0.12.14, local=0.12.12
ros2pkg                                   : latest=0.18.14, local=0.18.12
std_msgs                                  : latest=4.9.0, local=4.9.0
ros2topic                                 : latest=0.18.14, local=0.18.12
osrf_pycommon                             : latest=2.1.6, local=2.1.6
rcl_yaml_param_parser                     : latest=5.3.10, local=5.3.9
rpyutils                                  : latest=0.2.1, local=0.2.1
rmw_fastrtps_shared_cpp                   : latest=6.2.8, local=6.2.7
ament_cmake_xmllint                       : latest=0.12.14, local=0.12.12
rmw                                       : latest=6.1.2, local=6.1.2
ament_lint_cmake                          : latest=0.12.14, local=0.12.12
rclpy                                     : latest=3.3.17, local=3.3.16
rmw_fastrtps_cpp                          : latest=6.2.8, local=6.2.7
lifecycle_msgs                            : latest=1.2.2, local=1.2.1
rcl                                       : latest=5.3.10, local=5.3.9
ament_cmake_pytest                        : latest=1.3.12, local=1.3.11
rosidl_adapter                            : latest=3.1.7, local=3.1.6
std_srvs                                  : latest=4.9.0, local=4.9.0
ament_cmake_core                          : latest=1.3.12, local=1.3.11
ament_pep257                              : latest=0.12.14, local=0.12.12
ament_cmake_export_targets                : latest=1.3.12, local=1.3.11
fastrtps_cmake_module                     : latest=2.2.2, local=2.2.2
console_bridge_vendor                     : latest=1.4.1, local=1.4.1
ament_cmake_export_dependencies           : latest=1.3.12, local=1.3.11
rclcpp_action                             : latest=16.0.15, local=16.0.12
ament_cmake_uncrustify                    : latest=0.12.14, local=0.12.12
rcl_action                                : latest=5.3.10, local=5.3.9
ament_cmake_pep257                        : latest=0.12.14, local=0.12.12
builtin_interfaces                        : latest=1.2.2, local=1.2.1
rosidl_generator_py                       : latest=0.14.5, local=0.14.4
rcl_logging_interface                     : latest=2.3.1, local=2.3.1
pluginlib                                 : latest=5.1.0, local=5.1.0
rmw_implementation_cmake                  : latest=6.1.2, local=6.1.2
ros2param                                 : latest=0.18.14, local=0.18.12
ros2run                                   : latest=0.18.14, local=0.18.12
rosidl_typesupport_interface              : latest=3.1.7, local=3.1.6
common_interfaces                         : latest=4.9.0, local=4.9.0
tango_icons_vendor                        : latest=0.1.1, local=0.1.1
nav2_common                               : latest=1.1.18, local=1.1.18
rviz_common                               : latest=11.2.20, local=11.2.19
nav2_map_server                           : latest=1.1.18, local=1.1.18
map_msgs                                  : latest=2.1.0, local=2.1.0
rviz_rendering                            : latest=11.2.20, local=11.2.19
pcl_msgs                                  : latest=1.0.0, local=1.0.0
rqt_py_common                             : latest=1.1.7, local=1.1.7
bond                                      : latest=4.1.2, local=4.1.2
diagnostic_updater                        : latest=4.0.6, local=4.0.6
rviz2                                     : latest=11.2.20, local=11.2.19
rqt_topic                                 : latest=1.5.0, local=1.5.0
resource_retriever                        : latest=3.1.3, local=3.1.3
topic_tools                               : latest=1.1.1, local=1.1.1
libcurl_vendor                            : latest=3.1.3, local=3.1.3
interactive_markers                       : latest=2.3.2, local=2.3.2
nav2_lifecycle_manager                    : latest=1.1.18, local=1.1.18
smclib                                    : latest=4.1.2, local=4.1.2
rqt_gui                                   : latest=1.1.7, local=1.1.7
rviz_ogre_vendor                          : latest=11.2.20, local=11.2.19
rviz_assimp_vendor                        : latest=11.2.20, local=11.2.19
rqt_gui_py                                : latest=1.1.7, local=1.1.7
qt_gui                                    : latest=2.2.4, local=2.2.4
python_qt_binding                         : latest=1.1.2, local=1.1.2
pcl_conversions                           : latest=2.4.5, local=2.4.5
image_transport                           : latest=3.1.12, local=3.1.12
nav2_util                                 : latest=1.1.18, local=1.1.18
topic_tools_interfaces                    : latest=1.1.1, local=1.1.1
rqt_console                               : latest=2.0.3, local=2.0.3
laser_geometry                            : latest=2.4.0, local=2.4.0
bondcpp                                   : latest=4.1.2, local=4.1.2
rviz_default_plugins                      : latest=11.2.20, local=11.2.19
nav2_msgs                                 : latest=1.1.18, local=1.1.18
tf2_msgs                                  : latest=0.25.16, local=0.25.13
pybind11_vendor                           : latest=2.4.2, local=2.4.2
sqlite3_vendor                            : latest=0.15.15, local=0.15.14
rosbag2_cpp                               : latest=0.15.15, local=0.15.14
rosbag2_interfaces                        : latest=0.15.15, local=0.15.14
orocos_kdl_vendor                         : latest=0.2.5, local=0.2.5
tf2_eigen                                 : latest=0.25.16, local=0.25.13
zstd_vendor                               : latest=0.15.15, local=0.15.14
rosbag2_compression_zstd                  : latest=0.15.15, local=0.15.14
tf2_ros_py                                : latest=0.25.16, local=0.25.13
tf2_tools                                 : latest=0.25.16, local=0.25.13
rosbag2_compression                       : latest=0.15.15, local=0.15.14
rosbag2_storage                           : latest=0.15.15, local=0.15.14
keyboard_handler                          : latest=0.0.5, local=0.0.5
rosbag2_storage_default_plugins           : latest=0.15.15, local=0.15.14
tf2_bullet                                : latest=0.25.16, local=0.25.13
shared_queues_vendor                      : latest=0.15.15, local=0.15.14
tf2_sensor_msgs                           : latest=0.25.16, local=0.25.13
tf2_kdl                                   : latest=0.25.16, local=0.25.13
eigen3_cmake_module                       : latest=0.1.1, local=0.1.1
tf2                                       : latest=0.25.16, local=0.25.13
tinyxml_vendor                            : latest=0.8.3, local=0.8.3
urdf                                      : latest=2.6.1, local=2.6.1
sensor_msgs_py                            : latest=4.9.0, local=4.9.0
rosbag2_py                                : latest=0.15.15, local=0.15.14
tf2_ros                                   : latest=0.25.16, local=0.25.13
geometry2                                 : latest=0.25.16, local=0.25.13
tf2_eigen_kdl                             : latest=0.25.16, local=0.25.13
tf2_py                                    : latest=0.25.16, local=0.25.13
robot_state_publisher                     : latest=3.0.3, local=3.0.3
rosbag2                                   : latest=0.15.15, local=0.15.14
ros_base                                  : latest=0.10.0, local=0.10.0
urdf_parser_plugin                        : latest=2.6.1, local=2.6.1
rosbag2_transport                         : latest=0.15.15, local=0.15.14
message_filters                           : latest=4.3.8, local=4.3.7
yaml_cpp_vendor                           : latest=8.0.2, local=8.0.2
ros2bag                                   : latest=0.15.15, local=0.15.14
tf2_geometry_msgs                         : latest=0.25.16, local=0.25.13
python_orocos_kdl_vendor                  : latest=0.2.5, local=0.2.5
kdl_parser                                : latest=2.6.4, local=2.6.4

   PLATFORM INFORMATION
system           : Linux
platform info    : Linux-5.15.167.4-microsoft-standard-WSL2-x86_64-with-glibc2.35
release          : 5.15.167.4-microsoft-standard-WSL2
processor        : x86_64

   QOS COMPATIBILITY LIST
compatibility status    : No publisher/subscriber pairs found

   RMW MIDDLEWARE
middleware name    : rmw_fastrtps_cpp

   ROS 2 INFORMATION
distribution name      : humble
distribution type      : ros2
distribution status    : active
release platforms      : {'rhel': ['8'], 'ubuntu': ['jammy']}

   TOPIC LIST
topic               : none
publisher count     : 0
subscriber count    : 0
```
</details>


### Steps to reproduce issue

1. Create `some.launch.xml` bellow.
2. Run `ros2 launch some.launch.xml`
3. See console output.

## `some.launch.xml`

```xml
<?xml version="1.0" encoding="UTF-8"?>
<launch>
    <set_use_sim_time value="false" />
</launch>
```

### Expected behavior

The launch process exit successfully.

### Actual behavior

The launch exit with error as

```sh
$ ros2 launch some.launch.xml
[INFO] [launch]: All log files can be found below /root/.ros/log/2025-09-12-06-09-26-018286-7bf22331d729-37373
[INFO] [launch]: Default logging verbosity is set to INFO
[ERROR] [launch]: Caught exception in launch (see debug for traceback): Type 'Type.BOOL' and value '[<launch.substitutions.text_substitution.TextSubstitution object at 0x7f1fd4de9540>]' do not agree
```

### Additional information

The launch also failed when value was `False` or `$(eval False)` with different reasons.

## Use `False`

Use

```xml
<?xml version="1.0" encoding="UTF-8"?>
<launch>
    <set_use_sim_time value="False" />
</launch>
```

The output was

```sh
$ ros2 launch some.launch.xml
[INFO] [launch]: All log files can be found below /root/.ros/log/2025-09-12-06-18-45-931816-7bf22331d729-39134
[INFO] [launch]: Default logging verbosity is set to INFO
[ERROR] [launch]: Caught exception in launch (see debug for traceback): Type 'Type.BOOL' and value '[<launch.substitutions.text_substitution.TextSubstitution object at 0x7f1dcf4a4a30>]' do not agree
```

## Use `$(eval False)`

Use 

```xml
<?xml version="1.0" encoding="UTF-8"?>
<launch>
    <set_use_sim_time value="$(eval False)" />
</launch>
```

The output was

```sh
$ ros2 launch some.launch.xml
[INFO] [launch]: All log files can be found below /root/.ros/log/2025-09-12-06-22-54-038152-7bf22331d729-39936
[INFO] [launch]: Default logging verbosity is set to INFO
[ERROR] [launch]: Caught exception in launch (see debug for traceback): Type 'Type.BOOL' and value '[<launch.substitutions.python_expression.PythonExpression object at 0x7fd52f8e8730>]' do not agree
```

--- a/launch_ros/launch_ros/actions/set_use_sim_time.py
+++ b/launch_ros/launch_ros/actions/set_use_sim_time.py
@@ -12,12 +12,18 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+from typing import List
+from typing import Union
 
 from launch import Action
 from launch.frontend import Entity
 from launch.frontend import expose_action
 from launch.frontend import Parser
 from launch.launch_context import LaunchContext
+from launch.some_substitutions_type import SomeSubstitutionsType
+from launch.substitution import Substitution
+from launch.utilities.type_utils import normalize_typed_substitution
+from launch.utilities.type_utils import perform_typed_substitution
 
 from launch_ros.ros_adapters import get_ros_node
 
@@ -30,32 +36,33 @@ class SetUseSimTime(Action):
 
     def __init__(
         self,
-        value: bool,
+        value: Union[bool, SomeSubstitutionsType],
         **kwargs
     ) -> None:
         """Create a SetUseSimTime action."""
         super().__init__(**kwargs)
-        self.__value = value
+        self.__value = normalize_typed_substitution(value, bool)
 
     @classmethod
     def parse(cls, entity: Entity, parser: Parser):
         """Return `SetUseSimTime` action and kwargs for constructing it."""
         _, kwargs = super().parse(entity, parser)
-        kwargs['value'] = parser.parse_substitution(entity.get_attr('value'))
+        kwargs['value'] = parser.parse_if_substitutions(entity.get_attr('value', data_type=bool))
         return cls, kwargs
 
     @property
-    def value(self) -> bool:
+    def value(self) -> List[Substitution]:
         """Getter for value."""
         return self.__value
 
     def execute(self, context: LaunchContext):
         """Execute the action."""
+        value = perform_typed_substitution(context, self.value, bool)
         node = get_ros_node(context)
         param = Parameter(
             'use_sim_time',
             Parameter.Type.BOOL,
-            self.value
+            value,
         )
         node.set_parameters([param])
         if not node.get_parameter('use_sim_time').get_parameter_value().bool_value:

--- a/test_launch_ros/test/test_launch_ros/frontend/test_set_use_sim_timer_frontend.py
+++ b/test_launch_ros/test/test_launch_ros/frontend/test_set_use_sim_timer_frontend.py
@@ -1,0 +1,64 @@
+# Copyright 2025 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import io
+import textwrap
+
+from launch import LaunchService
+from launch.frontend import Parser
+from launch.utilities.type_utils import perform_typed_substitution
+from launch_ros.actions import SetUseSimTime
+import pytest
+
+
+yaml_file = textwrap.dedent(
+    r"""
+    launch:
+        - let:
+            name: sim_time
+            value: '1'
+        - set_use_sim_time:
+            value: $(var sim_time)
+    """
+)
+
+
+xml_file = textwrap.dedent(
+    r"""
+    <launch>
+        <let name="sim_time" value="1" />
+        <set_use_sim_time value="$(var sim_time)" />
+    </launch>
+    """
+)
+
+
+@pytest.mark.parametrize(
+    'file',
+    [
+        pytest.param(yaml_file, id='YAML'),
+        pytest.param(xml_file, id='XML'),
+    ],
+)
+def test_set_use_sim_timer(file):
+    root_entity, parser = Parser.load(io.StringIO(file))
+    ld = parser.parse_description(root_entity)
+    ls = LaunchService()
+    ls.include_launch_description(ld)
+    assert 0 == ls.run()
+
+    lc = ls.context
+    assert len(ld.entities) == 2
+    assert isinstance(ld.entities[1], SetUseSimTime)
+    assert perform_typed_substitution(lc, ld.entities[1].value, bool) is True

